### PR TITLE
Fix a bug where nn.functional._AllGather.backward produces wrong gradients

### LINKED ISSUE #120386: Wrong gradient value using `torch.distributed.nn.all_gather`
https://github.com/pytorch/pytorch/blob/5a114f72bf0aa93268ba51707a20f99b79ed048d/torch/distributed/nn/functional.py#L330-L341

It looks like using `_Reduce_Scatter.apply` (or `_AlltoAll.apply`) to compute gradient generates wrong (random) gradient. We can fix this by using `dist.reduce_scatter` instead.

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120582

Summary:
Fixes #120386

`_AllGather.backward` assumes that `_ReduceScatter` would always in-place update the output buffer. However, when the output buffer is non-contiguous, `_ReduceScatter` would allocate and return a different buffer, causing the gradient to be thrown away.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

### CODE REVIEW

**Review by XilunWu — APPROVED** [review]
LGTM! Thx for catching this bug in code. The other collectives seem to have no such issue.

**Comment by yifuwang** [comment]
> The other collectives seem to have no such issue.

Thanks for double checking!
